### PR TITLE
KIALI-1670 add support for groupBy options

### PR DESCRIPTION
**Pull request by jshaughn**
This is currently set to DNM, it needs to be demoed and determined to be useful.  It's also depended on some work in progress by @josejulio .

THIS PR REQUIRES SERVER PR: https://github.com/kiali/kiali/pull/659

### CODE REVIEW

**Comment by jmazzitelli** [comment]
I keep hitting F5 to refresh the graph - see what happens - flipping between grouping/not-grouping:

groupby ogv

**Comment by josejulio** [comment]
@jshaughn  try this for the `cose` layout

```
diff --git a/src/components/CytoscapeGraph/Layout/GroupCompoundLayout.ts b/src/components/CytoscapeGraph/Layout/GroupCompoundLayout.ts
index 6fa0404..f3e2246 100644
--- a/src/components/CytoscapeGraph/Layout/GroupCompoundLayout.ts
+++ b/src/components/CytoscapeGraph/Layout/GroupCompoundLayout.ts
@@ -67,6 +67,9 @@ class SyntheticEdgeGenerator {
   public getEdge(source: any, target: any) {
     const sourceId = this.normalizeToParent(source).id();
     const targetId = this.normalizeToParent(target).id();
+    if (sourceId === targetId) {
+      return false;
+    }
     const key = `${sourceId}->${targetId}`;
 
     if (this.generatedMap[key]) {
```

File: src/components/CytoscapeGraph/CytoscapeGraph.tsx [hunk 1]
```diff
@@ -356,7 +356,7 @@ export class CytoscapeGraph extends React.Component<CytoscapeGraphProps, Cytosca
 
     cy.startBatch();
 
-    // KIALI-1291 issue was caused because because some layouts (can't tell if all) do reuse the existing positions.
+    // KIALI-1291 issue was caused because some layouts (can't tell if all) do reuse the existing positions.
```
Comment (josejulio) [inline src/components/CytoscapeGraph/CytoscapeGraph.tsx]: thanks thanks

**Comment by jshaughn** [comment]
@josejulio That addition to GroupCompoundLayout seems to solve the issue I reported to you.  I have updated the PR with the change.

@jmazzitelli I can not replicate this so I'm not sure what to say about it.  I don't think it should block the PR.  There has been an update to the group layout code (see above in the comment).  Perhaps try again?

**Comment by jmazzitelli** [comment]
I can always make this happen. Here's another attempt (I just pulled the branches and rebuilt again):
Watch the reviews triangle:

groupby ogv

**Comment by josejulio** [comment]
> I can always make this happen. Here's another attempt (I just pulled the branches and rebuilt again):
> Watch the reviews triangle:

Do you get the same json or different jsons between refresh? Could you upload the json?

**Comment by jshaughn** [comment]
@jmazzitelli You are very annoying.  I think there is a real problem here in that the reviews service is actually *not* labeled with app=reviews.  And I think I'm making an assumption that a request for reviews service, handled by a workload with an app label, will mean that the service itself is labeled.  But this is not true and unfortunately it means that we're going to have to query istio/k8s for every service node, to check for the app label (only when groupBy=app).

**Comment by jmazzitelli** [comment]
> You are very annoying.

I think you meant to say that I am an exceptionally thorough reviewer with an attention to detail second-to-none. Thank you for the kind words.

**Comment by jmazzitelli** [comment]
groupby

UI console log says:

GET https://kiali-istio-system.192.168.1.19.nip.io/api/namespaces/graph?duration=60s&graphType=versionedApp&injectServiceNodes=true&groupBy=app&appenders=dead_node,sidecars_check,istio 500 (Internal Server Error)

Kiali server logs:

```
E1116 21:56:21.045063       1 graph.go:998] endpoints "ratings" not found: goroutine 145 [running]:
--
  | runtime/debug.Stack(0xc420df2220, 0x1d, 0xc420e52660)
  | /home/jmazzite/bin/go-install/go/src/runtime/debug/stack.go:24 +0xa7
  | github.com/kiali/kiali/handlers.handlePanic(0x13697c0, 0xc4202b8000)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/graph.go:998 +0xa6
  | panic(0x106b9a0, 0xc420e52660)
  | /home/jmazzite/bin/go-install/go/src/runtime/panic.go:502 +0x229
  | github.com/kiali/kiali/graph/appender.checkError(0x134f3c0, 0xc420630480)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/graph/appender/util.go:20 +0x9d
  | github.com/kiali/kiali/graph/appender.addLabels(0xc420e91b60, 0xc420e83600, 0xc420e592c0)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/graph/appender/istio_details.go:124 +0x157
  | github.com/kiali/kiali/graph/appender.IstioAppender.AppendGraph(0xc420e91b60, 0xc420e83600, 0xc420e592c0)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/graph/appender/istio_details.go:36 +0x82
  | github.com/kiali/kiali/handlers.buildNamespacesTrafficMap(0xc420f0c540, 0xc42066f140, 0x3, 0x4, 0xdf8475800, 0x100, 0xc4209862d0, 0x5bef3d05, 0x12401ae, 0x9, ...)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/graph.go:99 +0x28d
  | github.com/kiali/kiali/handlers.graphNamespaces(0x13697c0, 0xc4202b8000, 0xc420447300, 0xc420986240)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/graph.go:74 +0x13c
  | github.com/kiali/kiali/handlers.GraphNamespaces(0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/handlers/graph.go:63 +0xa3
  | net/http.HandlerFunc.ServeHTTP(0x12c3270, 0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:1947 +0x44
  | github.com/kiali/kiali/routing.metricHandler.func1(0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/routing/router.go:73 +0x97
  | net/http.HandlerFunc.ServeHTTP(0xc4203b0d20, 0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:1947 +0x44
  | github.com/kiali/kiali/config.AuthenticationHandler.func1(0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/config/authentication.go:31 +0x39b
  | net/http.HandlerFunc.ServeHTTP(0xc4203d4040, 0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:1947 +0x44
  | github.com/kiali/kiali/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc4203743f0, 0x13697c0, 0xc4202b8000, 0xc420447300)
  | /home/jmazzite/source/kiali/kiali/src/github.com/kiali/kiali/vendor/github.com/gorilla/mux/mux.go:162 +0xed
  | net/http.(*ServeMux).ServeHTTP(0x1aae200, 0x13697c0, 0xc4202b8000, 0xc420d0e300)
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:2337 +0x130
  | net/http.serverHandler.ServeHTTP(0xc4203cc0d0, 0x13697c0, 0xc4202b8000, 0xc420d0e300)
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:2694 +0xbc
  | net/http.(*conn).serve(0xc42036e140, 0x136a5c0, 0xc4208a4000)
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:1830 +0x651
  | created by net/http.(*Server).Serve
  | /home/jmazzite/bin/go-install/go/src/net/http/server.go:2795 +0x27b


```

**Comment by jshaughn** [comment]
@jmazzitelli So, I don't really know why you got that error.  Our underlying k8s API seems to have thrown an error trying to grab endpoints for the service.  It doesn't look like a problem in the new getServiceDefinitions() business layer call I added.  I don't think getting endpoints for a service should really throw an error in general.  Maybe @lucasponce or @jotak can comment, perhaps something to do with the latest cache stuff?

Anyway, what I did was to add protection in the appender so that if getting the service definition  fails we'll just log a debug message and continue.  We don't want the whole graph generation to fail.  The PR is updated.

**Comment by jshaughn** [comment]
Just waiting on the server PR to be merged.

**Pull request merged**